contracts/utils/Swap.sol

```diff
@@ -0,0 +1,463 @@
+    uint256 public immutable DOMAIN_CHAIN_ID;
+    bytes32 public immutable DOMAIN_SEPARATOR;
+
+    uint256 internal constant MAX_PERCENTAGE = 10000;
+    uint256 internal constant MAX_ERROR_COUNT = 8;
+    uint256 internal constant OTOKEN_DECIMALS = 8;
+
+    uint256 public offersCounter = 0;
+
+    mapping(uint256 => Offer) public swapOffers;
+
+    mapping(address => uint256) public referralFees;
+
+    /**
+     * @notice Double mapping of signers to nonce groups to nonce states
+     * @dev The nonce group is computed as nonce / 256, so each group of 256 sequential nonces uses the same key
+     * @dev The nonce states are encoded as 256 bits, for each nonce in the group 0 means available and 1 means used
+     */
+    mapping(address => mapping(uint256 => uint256)) internal _nonceGroups;
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    constructor() {
+        uint256 currentChainId = getChainId();
+        DOMAIN_CHAIN_ID = currentChainId;
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                currentChainId,
+                this
+            )
+        );
+    }
+
+    /************************************************
+     *  SETTER
+     ***********************************************/
+
+    /**
+     * @notice Sets the referral fee for a specific referrer
+     * @param referrer is the address of the referrer
+     * @param fee is the fee in percent in 2 decimals
+     */
+    function setFee(address referrer, uint256 fee) external onlyOwner {
+        require(referrer != address(0), "Referrer cannot be the zero address");
+        require(fee < MAX_PERCENTAGE, "Fee exceeds maximum");
+
+        referralFees[referrer] = fee;
+    }
+
+    /************************************************
+     *  OFFER CREATION AND SETTLEMENT
+     ***********************************************/
+
+    /**
+     * @notice Create a new offer available for swap
+     * @param oToken token offered by seller
+     * @param biddingToken token asked by seller
+     * @param minPrice minimum price of oToken denominated in biddingToken
+     * @param minBidSize minimum amount of oToken requested in a single bid
+     * @param totalSize amount of oToken offered by seller
+     */
+    function createOffer(
+        address oToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(oToken != address(0), "oToken cannot be the zero address");
+        require(
+            biddingToken != address(0),
+            "BiddingToken cannot be the zero address"
+        );
+        require(minPrice > 0, "MinPrice must be larger than zero");
+        require(minBidSize > 0, "MinBidSize must be larger than zero");
```

Comment:
We can change this to be a check that `require(minBidSize <= totalSize)`, so that we don't accidentally void the Offering because no one can bid.